src/output.rs

```diff
@@ -1,693 +1,674 @@
-//! Types and functions related to graphical outputs
-//!
-//! This modules provides two main elements. The first is the
-//! [`OutputHandler`](struct.OutputHandler.html) type, which is a
-//! [`MultiGlobalHandler`](../environment/trait.MultiGlobalHandler.html) for
-//! use with the [`init_environment!`](../macro.init_environment.html) macro. It is automatically
-//! included if you use the [`new_default_environment!`](../macro.new_default_environment.html).
-//!
-//! The second is the [`with_output_info`](fn.with_output_info.html) with allows you to
-//! access the information associated to this output, as an [`OutputInfo`](struct.OutputInfo.html).
-
 use std::{
-    cell::RefCell,
-    fmt,
-    rc::{self, Rc},
-    sync::{self, Arc, Mutex},
+    fmt::{self, Display, Formatter},
+    mem,
+    sync::{Arc, Mutex},
 };
 
 use wayland_client::{
-    protocol::{
-        wl_output::{self, Event, WlOutput},
-        wl_registry,
-    },
-    Attached, DispatchData, Main,
+    protocol::wl_output::{self, Subpixel, Transform},
+    ConnectionHandle, DataInit, DelegateDispatch, DelegateDispatchBase, Dispatch, Proxy,
+    QueueHandle, WEnum,
 };
-
 use wayland_protocols::unstable::xdg_output::v1::client::{
-    zxdg_output_manager_v1::ZxdgOutputManagerV1,
-    zxdg_output_v1::{self, ZxdgOutputV1},
+    zxdg_output_manager_v1::{self, ZxdgOutputManagerV1},
+    zxdg_output_v1,
 };
 
-pub use wayland_client::protocol::wl_output::{Subpixel, Transform};
+use crate::registry::{RegistryHandle, RegistryHandler};
 
-/// A possible mode for an output
-#[derive(Copy, Clone, Debug)]
-pub struct Mode {
-    /// Number of pixels of this mode in format `(width, height)`
+pub trait OutputHandler {
+    /// A new output has been advertised.
+    fn new_output(&mut self, info: OutputInfo);
+
+    /// An existing output has changed.
+    fn update_output(&mut self, info: OutputInfo);
+
+    /// An output is no longer advertised.
     ///
-    /// for example `(1920, 1080)`
-    pub dimensions: (i32, i32),
-    /// Refresh rate for this mode, in mHz
-    pub refresh_rate: i32,
-    /// Whether this is the current mode for this output
-    pub is_current: bool,
-    /// Whether this is the preferred mode for this output
-    pub is_preferred: bool,
+    /// The info passed to this function was the state of the output before destruction.
+    fn output_destroyed(&mut self, info: OutputInfo);
 }
```

Comment:
These methods need to pass the `WlOutput` object as well, of the user will not be able to do anything meaningful with them.